api-utils/JumpCloud_API_Go_Examples/DedupeSystemUsersUnixUidGuid/dedupeSystemUsersUnixUidGuid.go

```diff
@@ -0,0 +1,145 @@
+	flag.StringVar(&url, "url", URLBase, "Alternative Jumpcloud API URL (optional)")
+	flag.BoolVar(&dryRun, "dryRun", true, "If false, send requests. If true, print out actions")
+	flag.Parse()
+
+	if apiKey == "" {
+		fmt.Printf("API Key must be provided.")
+		return
+	}
+
+	// Attach to JumpCloud
+	api = jcapi.NewJCAPI(apiKey, url)
+
+	// Fetch all users in JumpCloud
+	// Send false, because we do not want tags
+	fmt.Println("Retreiving users from JumpCloud. Depending on the number of users in your organization this may take awhile")
+	userList, err := api.GetSystemUsers(false)
+
+	if err != nil {
+		fmt.Printf("Could not read system users, err='%s'\n", err)
+		return
+	}
+	fmt.Println("Users retrieved. Finding users with duplicate unix_uid and unix_guid values")
+
+	dupeUsers, uidValues, seenUids, seenGuids := findDupeUsers(userList)
+
+	if len(dupeUsers) == 0 {
+		fmt.Println("No users have duplicate values! Nothing to do here")
+		return
+	}
+
+	reader := bufio.NewReader(os.Stdin)
+	fmt.Println("Now we can help you fix your duplicate values")
+	fmt.Println("---------------------------------------------")
+	fmt.Println("To fix the duplicate values, we are going to iterate through all users with duplicate values")
+	fmt.Println("And assign them the next available matching value for unix_uid and unix_guid")
+	fmt.Println("To continue with updating user values, type CONTINUE, otherwise type QUIT to stop script")
+	if dryRun {
+		fmt.Println("DRY RUN IS ON. USERS WILL NOT BE UPDATED. SET --dryRun=false TO UPDATE USER DATA")
+	}
+
+Loop:
+	for {
+		fmt.Print("-> ")
+		option, _ := reader.ReadString('\n')
+		option = strings.Replace(option, "\n", "", -1)
+		switch option {
+		case "CONTINUE":
+			break Loop
+		case "QUIT":
+			return
+		default:
+			fmt.Println("Enter a valid option to continue")
+		}
+	}
+	setValue := uidValues[0]
+	for _, user := range dupeUsers {
+		fmt.Printf("Updating user: %s\n", user.Id)
+		setValue = findNextAvaiableValue(setValue, seenUids, seenGuids)
+		if dryRun {
+			fmt.Printf("Previous values for user: %s unix_uid: %s unix_guid %s\n", user.Id, user.Uid, user.Gid)
+			fmt.Printf("Would update user: %s with unix_uid: %d unix_guid: %d\n", user.Id, setValue, setValue)
+		} else {
+			user.Uid = strconv.Itoa(setValue)
+			user.Gid = strconv.Itoa(setValue)
+			_, err := api.AddUpdateUser(jcapi.Update, user)
+			if err != nil {
+				fmt.Println("Error updating user: %s. Bailing", err)
+				return
+			}
+		}
+		fmt.Printf("User: %s updated\n", user.Id)
+	}
+	fmt.Printf("All users updated!\n")
+	return
+}
+
+func findDupeUsers(userList []jcapi.JCUser) ([]jcapi.JCUser, []int, map[int]bool, map[int]bool) {
+	seenUids := make(map[int]bool)
+	seenGuids := make(map[int]bool)
+	dupeUsers := make([]jcapi.JCUser, 0)
+	uidValues := make([]int, 0)
+	guidValues := make([]int, 0)
+
+	dupeUid := false
+	dupeGuid := false
+	for _, user := range userList {
+		uidInt, _ := strconv.Atoi(user.Uid)
```

Comment:
This treats all non-integer values as 0 (and therefore dups of each other). We recently learned that there are around 2k systemusers in prd with `unix_uid = null` (see https://jumpcloud.atlassian.net/browse/SJ-352).  We'll have to think about this and probably discuss with @ppg but my inclination would be to consider `null` values to be valid (e.g. windows-only systemusers don't need a `uid`). In that case, we would want to omit `null` values from our dup sets.